pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java

```diff
@@ -0,0 +1,175 @@
+
+    for (int i = 0; i < length; i++) {
+      try {
+        HyperLogLog value = HllUtil.convertStringToHll(values[i]);
+        hll.addAll(value);
+      } catch (Exception e) {
+        Utils.rethrowException(e);
+      }
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * While the interface allows for variable number of valueArrays, we do not support
+   * multiple columns within one aggregation function right now.
+   *
+   * @param length
+   * @param groupKeys
+   * @param resultHolder
+   * @param valueArray
+   */
+  @Override
+  public void aggregateGroupBySV(int length, int[] groupKeys, GroupByResultHolder resultHolder, Object... valueArray) {
+    Preconditions.checkArgument(valueArray.length == 1);
+    Preconditions.checkArgument(valueArray[0] instanceof String[]);
+    final String[] values = (String[]) valueArray[0];
+    Preconditions.checkState(length <= values.length);
+
+    for (int i = 0; i < length; i++) {
+      int groupKey = groupKeys[i];
+      HyperLogLog hll = resultHolder.getResult(groupKey);
+      if (hll == null) {
+        hll = new HyperLogLog(hllLog2m);
+        resultHolder.setValueForKey(groupKey, hll);
+      }
+      try {
+        HyperLogLog value = HllUtil.convertStringToHll(values[i]);
+        hll.addAll(value);
+      } catch (Exception e) {
+        Utils.rethrowException(e);
+      }
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param length
+   * @param docIdToGroupKeys
+   * @param resultHolder
+   * @param valueArray
+   */
+  @Override
+  public void aggregateGroupByMV(int length, int[][] docIdToGroupKeys, GroupByResultHolder resultHolder,
+      Object... valueArray) {
+    Preconditions.checkArgument(valueArray.length == 1);
+    Preconditions.checkArgument(valueArray[0] instanceof String[]);
+    final String[] values = (String[]) valueArray[0];
+    Preconditions.checkState(length <= values.length);
+
+    for (int i = 0; i < length; i++) {
+      for (int groupKey : docIdToGroupKeys[i]) {
+        HyperLogLog hll = resultHolder.getResult(groupKey);
+        if (hll == null) {
+          hll = new HyperLogLog(hllLog2m);
+          resultHolder.setValueForKey(groupKey, hll);
+        }
+        try {
+          HyperLogLog value = HllUtil.convertStringToHll(values[i]);
+          hll.addAll(value);
+        } catch (Exception e) {
+          Utils.rethrowException(e);
+        }
+      }
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @return
+   */
+  @Override
+  public double getDefaultValue() {
+    throw new RuntimeException("Unsupported method getDefaultValue() for class " + getClass().getName());
+  }
+
+  /**
+   * {@inheritDoc}
+   * @return
+   */
+  @Override
+  public ResultDataType getResultDataType() {
+    return RESULT_DATA_TYPE;
+  }
+
+  @Override
+  public String getName() {
+    return FUNCTION_NAME;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param combinedResult
+   * @return
+   */
+  @Override
+  public Double reduce(List<Object> combinedResult) {
```

Comment:
The reduce part of this new aggregation function is never got called. I think the transition of old aggregation API to new API is not fully completed. Currently the reduce is supported by old API and I think it should work fine for multiple segments. (That's part of the reason we need to keep the old API in system)